### PR TITLE
chore(flush on shutdown): validate s3 sink flushes

### PR DESCRIPTION
**Pull request by DominicBurkart**
Adding regression test for related issue: https://github.com/vectordotdev/vector/issues/11405

### CODE REVIEW

File: src/sinks/aws_s3/integration_tests.rs [hunk 1]
```diff
@@ -406,6 +406,52 @@
...
+    assert_eq!(keys.len(), 1);
+
+    let mut response_lines: Vec<String> = Vec::new();
+    let mut key_stream = stream::iter(keys);
+    while let Some(key) = key_stream.next().await {
+        let obj = get_object(&bucket, key).await;
+        response_lines.append(&mut get_lines(obj).await);
+    }
+
+    assert_eq!(lines, response_lines); // if all events are recieved, and lines.len() < batch size, then a flush was performed.
```
[the remark this answers is not in the document]
Comment (DominicBurkart) [reply src/sinks/aws_s3/integration_tests.rs]: force pushing to erase evidence of my bad spelling 😶

File: src/sinks/aws_s3/integration_tests.rs [hunk 2]
```diff
@@ -406,6 +406,51 @@ async fn s3_healthchecks_invalid_bucket() {
...
+            3
+        };
+        e.insert("i", i.to_string());
+        Event::from(e)
+    });
+
+    // Here, we validate that the s3 sink flushes when its source stream is exhausted
+    // by giving it a number of inputs less than the batch size, verifying that the
+    // outputs for the in-flight batch are flushed.
+    run_and_assert_sink_compliance(sink, stream::iter(events), &AWS_SINK_TAGS).await;
```
Comment (lukesteensen) [inline src/sinks/aws_s3/integration_tests.rs:439]: The only assertion that I think _might_ be missing here is one that the sink is not waiting for a batch timeout when shutting down.

Most of the time, batching configs for things like S3 will specify a trigger to flush at a minimum number of events OR a maximum amount of time (e.g. every 1k events or every 5 mins, whichever comes first). This test would probably take a long time if that were happening, but it'd be nice to make it explicit that we are not waiting. A simple way to do that would probably be to explicitly configure that timeout on the sink, and then add a timeout to this statement that is significantly shorter.

Comment (DominicBurkart) [reply src/sinks/aws_s3/integration_tests.rs]: Great point! Implemented here: https://github.com/vectordotdev/vector/pull/17667/commits/af35d9abcb154098ce198330fa4bf331c3c0ee52

**Review by lukesteensen — APPROVED** [review]
Looks good to me!

File: src/sinks/aws_s3/integration_tests.rs [hunk 3]
```diff
@@ -406,6 +406,80 @@ async fn s3_healthchecks_invalid_bucket() {
...
+    };
+    let prefix = config.key_prefix.clone();
+    let service = config.create_service(&cx.globals.proxy).await.unwrap();
+    let sink = config.build_processor(service).unwrap();
+
+    let (lines, _events) = random_lines_with_stream(100, 2, None); // only generate two events (less than batch size)
+
+    let events = lines.clone().into_iter().enumerate().map(|(i, line)| {
+        let mut e = LogEvent::from(line);
+        let i = if i < 10 {
```
Comment (jszwedko) [inline src/sinks/aws_s3/integration_tests.rs:447]: Won't `i` always be less than `10` since we only generate two events?

**Pull request merged**